File: backend/src/main/java/com/mini/credit/service/RecetteTerrainService.java
package com.mini.credit.service;

import com.mini.credit.dto.referentiel.CreateRecetteTerrainRequest;
import com.mini.credit.dto.referentiel.RecetteTerrainResponse;
import com.mini.credit.dto.referentiel.UpdateRecetteTerrainRequest;
import com.mini.credit.dto.referentiel.ValidateRecetteTerrainRequest;
import com.mini.credit.enums.RecetteStatut;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDate;
import java.util.List;

/**
 * Service pour gestion des recettes terrain journalières
 * Contient la logique métier pour création, validation, workflow
 */
public interface RecetteTerrainService {

    // ========== CRUD ==========

    /**
     * Crée une nouvelle recette terrain (statut BROUILLON)
     * Validations:
     * - Agent existe et actif
     * - Site existe et actif
     * - Agent et Site même agence
     * - Agent affecté au site
     * - Pas de doublon (agent + date)
     */
    RecetteTerrainResponse create(CreateRecetteTerrainRequest request);

    /**
     * Modifie une recette existante (seulement si BROUILLON)
     * Recalcule les écarts
     */
    RecetteTerrainResponse update(Long id, UpdateRecetteTerrainRequest request);

    /**
     * Récupère une recette par ID
     */
    RecetteTerrainResponse getById(Long id);

    /**
     * Récupère toutes les recettes actives
     */
    List<RecetteTerrainResponse> getAll();

    /**
     * Soft-delete d'une recette (actif = false)
     * Possible seulement si BROUILLON
     */
    void delete(Long id);

    // ========== RECHERCHES ==========

    /**
     * Récupère les recettes d'un agent terrain
     */
    List<RecetteTerrainResponse> getByAgentTerrain(Long agentTerrainId);

    /**
     * Récupère les recettes de l'agent terrain connecté.
     */
    List<RecetteTerrainResponse> getMine();

    /**
     * Récupère les recettes d'un site
     */
    List<RecetteTerrainResponse> getBySite(Long siteId);

    /**
     * Récupère les recettes d'une date donnée
     */
    List<RecetteTerrainResponse> getByDateRecette(LocalDate date);

    /**
     * Récupère les recettes par statut
     */
    List<RecetteTerrainResponse> getByStatut(RecetteStatut statut);

    /**
     * Récupère les recettes d'un site par statut
     */
    List<RecetteTerrainResponse> getBySiteAndStatut(Long siteId, RecetteStatut statut);

    /**
     * Récupère les recettes d'une plage de dates
     */
    List<RecetteTerrainResponse> getByDateRange(LocalDate debut, LocalDate fin);

    /**
     * Récupère les recettes en attente de validation pour un site
     */
    List<RecetteTerrainResponse> getEnAttenteValidation(Long siteId);

    // ========== WORKFLOW ==========

    /**
     * Soumet une recette pour validation (BROUILLON → SOUMISE)
     * Vérifications:
     * - Statut = BROUILLON
     * - Tous champs obligatoires présents (isComplete)
     * - Montants cohérents
     */
    RecetteTerrainResponse soumettre(Long id);

    /**
     * Valide ou rejette une recette soumise (SOUMISE → VALIDEE/REJETEE)
     * Réservé aux responsables site et contrôleurs
     * Si REJETEE: motif requis
     */
    RecetteTerrainResponse valider(Long id, ValidateRecetteTerrainRequest request);

    /**
     * Rejette une recette (raccourci pour valider avec REJETEE)
     */
    RecetteTerrainResponse rejeter(Long id, String motifRejet, Long validePar);

    // ========== MÉTIERS ==========

    /**
     * Calcule/recalcule les écarts (excédent ou manquant)
     * Appelé automatiquement à la création/modification
     */
    RecetteTerrainResponse calculerEcarts(Long id);

    /**
     * Récupère le total collecté pour un agent à une date donnée
     */
    java.math.BigDecimal getTotalCollecteByAgentAndDate(Long agentTerrainId, LocalDate date);

    /**
     * Valide qu'un agent terrain existe et est actif
     */
    void validerAgent(Long agentTerrainId);

    /**
     * Valide qu'un site existe et est actif
     */
    void validerSite(Long siteId);

    /**
     * Valide que l'agent est affecté au site
     */
    void validerAgentAffecteAuSite(Long agentTerrainId, Long siteId);

    /**
     * Valide que crédit et site sont même agence
     */
    void validerMemeAgence(Long agentTerrainId, Long siteId);

    // ========== PAGINATION ==========

    /**
     * Récupère toutes les recettes paginées
     */
    Page<RecetteTerrainResponse> getAllPaginated(Pageable pageable);

    /**
     * Recherche et filtre les recettes avec pagination
     * Paramètres optionnels: statut, siteId, agentTerrainId, dateDebut, dateFin
     * @param pageable pagination parameters (page, size, sort)
     * @param statut filter by statut (nullable)
     * @param siteId filter by site (nullable)
     * @param agentTerrainId filter by agent (nullable)
     * @param dateDebut filter by start date (nullable)
     * @param dateFin filter by end date (nullable)
     */
    Page<RecetteTerrainResponse> searchAndFilter(
        Pageable pageable,
        RecetteStatut statut,
        Long siteId,
        Long agentTerrainId,
        LocalDate dateDebut,
        LocalDate dateFin);
}

File: frontend/src/app/features/recettes/services/recette-terrain.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { API_BASE_URL } from '../../../core/services/api.config';

import {
  RecetteTerrainResponse,
  CreateRecetteTerrainRequest,
  UpdateRecetteTerrainRequest,
  ValidateRecetteTerrainRequest,
  RecetteStatut
} from '../models';

/**
 * Page Response DTO from Spring Data
 */
export interface PageResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  currentPage: number;
  pageSize: number;
  last: boolean;
  first: boolean;
  empty: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class RecetteTerrainService {

  private http = inject(HttpClient);
  private readonly baseUrl = `${API_BASE_URL}/recettes-terrain`;

  // === CRUD ===

  getAll(): Observable<RecetteTerrainResponse[]> {
    return this.http.get<RecetteTerrainResponse[]>(this.baseUrl);
  }

  getById(id: number): Observable<RecetteTerrainResponse> {
    return this.http.get<RecetteTerrainResponse>(`${this.baseUrl}/${id}`);
  }

  create(request: CreateRecetteTerrainRequest): Observable<RecetteTerrainResponse> {
    return this.http.post<RecetteTerrainResponse>(this.baseUrl, request);
  }

  update(id: number, request: UpdateRecetteTerrainRequest): Observable<RecetteTerrainResponse> {
    return this.http.put<RecetteTerrainResponse>(`${this.baseUrl}/${id}`, request);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  // === Recherches ===

  getByAgentTerrain(agentId: number): Observable<RecetteTerrainResponse[]> {
    return this.http.get<RecetteTerrainResponse[]>(`${this.baseUrl}/agent/${agentId}`);
  }

  getMine(): Observable<RecetteTerrainResponse[]> {
    return this.http.get<RecetteTerrainResponse[]>(`${this.baseUrl}/me`);
  }

  getBySite(siteId: number): Observable<RecetteTerrainResponse[]> {
    return this.http.get<RecetteTerrainResponse[]>(`${this.baseUrl}/site/${siteId}`);
  }

  getByStatut(statut: RecetteStatut): Observable<RecetteTerrainResponse[]> {
    return this.http.get<RecetteTerrainResponse[]>(`${this.baseUrl}/statut/${statut}`);
  }

  // === Workflow ===

  soumettre(id: number): Observable<RecetteTerrainResponse> {
    return this.http.post<RecetteTerrainResponse>(`${this.baseUrl}/${id}/soumettre`, {});
  }

  valider(id: number, request: ValidateRecetteTerrainRequest): Observable<RecetteTerrainResponse> {
    return this.http.post<RecetteTerrainResponse>(`${this.baseUrl}/${id}/valider`, request);
  }

  rejeter(id: number, motif: string, validePar: number): Observable<RecetteTerrainResponse> {
    return this.http.post<RecetteTerrainResponse>(
      `${this.baseUrl}/${id}/rejeter`,
      {},
      { params: { motif, validePar: validePar.toString() } }
    );
  }

  // === Métiers ===

  calculerEcarts(id: number): Observable<{ excedent: number; manquant: number; isBalanced: boolean }> {
    return this.http.get<{ excedent: number; manquant: number; isBalanced: boolean }>(
      `${this.baseUrl}/${id}/ecarts`
    );
  }

  getTotalCollecte(agentId: number, dateRecette: string): Observable<number> {
    return this.http.get<number>(
      `${this.baseUrl}/agent/${agentId}/date/${dateRecette}/total`
    );
  }

  /**
   * PHASE 6B.2: Déclenche manuellement la génération des opérations pour une recette
   */
  generateOperations(id: number): Observable<RecetteTerrainResponse> {
    return this.http.post<RecetteTerrainResponse>(
      `${this.baseUrl}/${id}/generate-operations`,
      {}
    );
  }

  /**
   * Récupère les recettes paginées avec filtres optionnels
   * @param page page number (0-indexed)
   * @param pageSize items per page
   * @param sortBy sort field (default: dateRecette)
   * @param sortDir sort direction (asc, desc)
   * @param statut optional status filter
   * @param siteId optional site filter
   * @param agentTerrainId optional agent filter
   * @param dateDebut optional start date filter
   * @param dateFin optional end date filter
   */
  searchAndFilterPaginated(
    page: number = 0,
    pageSize: number = 10,
    sortBy: string = 'dateRecette',
    sortDir: 'asc' | 'desc' = 'desc',
    statut?: string,
    siteId?: number,
    agentTerrainId?: number,
    dateDebut?: string,
    dateFin?: string
  ): Observable<PageResponse<RecetteTerrainResponse>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', pageSize.toString())
      .set('sort', `${sortBy},${sortDir}`);

    if (statut) params = params.set('statut', statut);
    if (siteId) params = params.set('siteId', siteId.toString());
    if (agentTerrainId) params = params.set('agentTerrainId', agentTerrainId.toString());
    if (dateDebut) params = params.set('dateDebut', dateDebut);
    if (dateFin) params = params.set('dateFin', dateFin);

    return this.http.get<PageResponse<RecetteTerrainResponse>>(
      `${this.baseUrl}/paginated`,
      { params }
    );
  }
}
